README.md
# ribbon_safe

A Ribbon decoration on a widget.
![](screenshot/s1.png)

use a Ribbon to decorate a Widget such as container,card...etc.

no clip apply to ribbon title,so check the title length

Installing:

In your pubspec.yaml

dependencies:

ribbon_safe: ^0.0.1

Example Usage:

import 'package:ribbon_safe/ribbon_safe.dart';

Ribbon(

nearLength: nearLength,

farLength: farLength,

title: 'New!',

titleStyle: TextStyle(

color: Colors.greenAccent,

fontSize: 18,

fontWeight: FontWeight.bold),

color: Colors.redAccent,

location: location,

child: )



<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

TODO: Put a short description of the package here that helps potential users
know whether this package might be useful for them.

## Features

TODO: List what your package can do. Maybe include images, gifs, or videos.

## Getting started

This project is a starting point for a Dart
[package](https://flutter.io/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our
[online documentation](https://flutter.io/docs), which offers tutorials,
samples, guidance on mobile development, and a full API reference.



CHANGELOG.md
## 0.0.1

* TODO: Describe initial release.

## 0.0.2
* TODO: Implementation Upgrade

test/ribbon_safe_test.dart
import 'package:flutter/material.dart';
import 'package:ribbon_safe/ribbon_safe.dart';

void main() {
  runApp(MaterialApp(
    title: "ribbon demo",
    theme: ThemeData.light(),
    home: DemoPage(),
    // debugShowMaterialGrid:true,
  ));
}

class DemoPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Ribbon(
            nearLength: 10,
            farLength: 20,
            title: 'ribbon',
            color: Colors.blueAccent,child: Container(color: Colors.green,width:300,height:150,),),
      ),
    );
  }
}


lib/ribbon_safe.dart
library ribbon_safe;

import 'package:flutter/material.dart';
import 'dart:math' as math;

const TextStyle _kTextStyle = TextStyle(
    color: Color(0xFFFFFFFF),
    fontSize: 12,
    fontWeight: FontWeight.w900,
    height: 1.0
);



enum RibbonLocation {
  topStart,
  topEnd,
  bottomStart,
  bottomEnd,
}

class Ribbon extends StatelessWidget {
  final double? nearLength;
  final double? farLength;
  final String? title;
  final Color? color;
  final TextStyle? titleStyle;
  final RibbonLocation? location;
  final Widget? child;
  const Ribbon(
      {Key? key,
        @required this.nearLength,
        @required this.farLength,
        @required this.title,
        this.titleStyle = _kTextStyle,
        this.color = Colors.white,
        this.location = RibbonLocation.topStart,
        this.child})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
        foregroundPainter: _RibbonPainter(
          nearLength: nearLength!,
          farLength: farLength!,
          title: title!,
          titleStyle: titleStyle!,
          color: color!,
          location: location!,
        ),
        child: child);
  }
}

class _RibbonPainter extends CustomPainter {
  double? nearLength;
  double? farLength;
  final String? title;
  final Color? color;
  final TextStyle? titleStyle;
  final RibbonLocation? location;
  bool? initialized = false;
  TextPainter? textPainter;
  Paint? paintRibbon;
  Path? pathRibbon;
  double? rotateRibbon;
  Offset? offsetRibbon;
  Offset? offsetTitle;
  Paint? paintShadow;

  static const BoxShadow _shadow = BoxShadow(
    color: Color(0x7F000000),
    blurRadius: 6.0,
  );
  _RibbonPainter(
      {this.nearLength,
        this.farLength,
        this.title,
        this.titleStyle,
        this.color,
        this.location});
  @override
  void paint(Canvas canvas, Size size) {
    if (!initialized!) _initializ(size);
    // canvas.drawPath(pathRibbon, paintShadow);
    canvas
      ..drawShadow(pathRibbon!, Color(0x7F000000), 2.0, true)
      ..drawPath(pathRibbon!, paintRibbon!)
    // canvas
      ..translate(offsetRibbon!.dx, offsetRibbon!.dy)
      ..rotate(rotateRibbon!);
    // ..clipPath(pathRibbon);
    textPainter!.paint(canvas, offsetTitle!);
  }

  @override
  bool shouldRepaint(_RibbonPainter oldDelegate) {
    return title != oldDelegate.title ||
        nearLength != oldDelegate.nearLength ||
        farLength != oldDelegate.farLength ||
        color != oldDelegate.color ||
        location != oldDelegate.location;
  }

  void _initializ(Size size) {
    initialized = true;
    if (nearLength! > farLength!) {
      double temp = farLength!;
      farLength = nearLength!;
      nearLength = temp;
    }
    if (farLength! > size.width) farLength = size.width;
    TextSpan span =
    TextSpan(style: titleStyle, text: title);
    textPainter = TextPainter(
        text: span,
        textAlign: TextAlign.center,
        textDirection: TextDirection.ltr);
    textPainter!.layout();
    paintRibbon = Paint()
      ..color = color!
      ..style = PaintingStyle.fill;
    offsetTitle = Offset(-textPainter!.width / 2, -textPainter!.height / 2);
    rotateRibbon = _rotation;
    pathRibbon = _ribbonPath(size);
    paintShadow = _shadow.toPaint();
  }

  Path _ribbonPath(Size size) {
    Path path = Path();
    List<Offset> vec = [];
    if (size.width <= size.height) {
      switch (location) {
        case RibbonLocation.topStart:
          path.moveTo(nearLength!, 0);
          vec.add(Offset(nearLength!, 0));
          path.lineTo(farLength!, 0);
          vec.add(Offset(farLength!, 0));
          path.lineTo(0, farLength!);
          vec.add(Offset(0, farLength!));
          path.lineTo(0, nearLength!);
          vec.add(Offset(0, nearLength!));
          break;
        case RibbonLocation.topEnd:
          path.moveTo(size.width - nearLength!, 0);
          vec.add(Offset(size.width - nearLength!, 0));
          path.lineTo(size.width - farLength!, 0);
          vec.add(Offset(size.width - farLength!, 0));
          path.lineTo(size.width, farLength!);
          vec.add(Offset(size.width, farLength!));
          path.lineTo(size.width, nearLength!);
          vec.add(Offset(size.width, nearLength!));
          break;
        case RibbonLocation.bottomStart:
          path.moveTo(0, size.height - nearLength!);
          vec.add(Offset(0, size.height - nearLength!));
          path.lineTo(0, size.height - farLength!);
          vec.add(Offset(0, size.height - farLength!));
          path.lineTo(farLength!, size.height);
          vec.add(Offset(farLength!, size.height));
          path.lineTo(nearLength!, size.height);
          vec.add(Offset(nearLength!, size.height));
          break;
        case RibbonLocation.bottomEnd:
          path.moveTo(size.width - nearLength!, size.height);
          vec.add(Offset(size.width - nearLength!, size.height));
          path.lineTo(size.width - farLength!, size.height);
          vec.add(Offset(size.width - farLength!, size.height));
          path.lineTo(size.width, size.height - farLength!);
          vec.add(Offset(size.width, size.height - farLength!));
          path.lineTo(size.width, size.height - nearLength!);
          vec.add(Offset(size.width, size.height - nearLength!));
          break;
      }
    } else {
      switch (location) {
        case RibbonLocation.topStart:
          path.moveTo(nearLength!, 0);
          vec.add(Offset(nearLength!, 0));
          path.lineTo(farLength!, 0);
          vec.add(Offset(farLength!, 0));
          if (farLength! <= size.height) {
            path.lineTo(0, farLength!);
            vec.add(Offset(0, farLength!));
            path.lineTo(0, nearLength!);
            vec.add(Offset(0, nearLength!));
          } else {
            path.lineTo(farLength! - size.height, size.height);
            vec.add(Offset(farLength! - size.height, size.height));
            if (nearLength! <= size.height) {
              path.lineTo(0, size.height);
              vec.add(Offset(0, size.height));
              path.lineTo(0, nearLength!);
              vec.add(Offset(0, nearLength!));
            } else {
              path.lineTo(nearLength! - size.height, size.height);
              vec.add(Offset(nearLength! - size.height, size.height));
            }
          }
          break;
        case RibbonLocation.topEnd:
          path.moveTo(size.width - nearLength!, 0);
          vec.add(Offset(size.width - nearLength!, 0));
          path.lineTo(size.width - farLength!, 0);
          vec.add(Offset(size.width - farLength!, 0));
          if (farLength! <= size.height) {
            path.lineTo(size.width, farLength!);
            vec.add(Offset(size.width, farLength!));
            path.lineTo(size.width, nearLength!);
            vec.add(Offset(size.width, nearLength!));
          } else {
            path.lineTo(size.width - (farLength! - size.height), size.height);
            vec.add(
                Offset(size.width - (farLength! - size.height), size.height));
            if (nearLength! <= size.height) {
              path.lineTo(size.width, size.height);
              vec.add(Offset(size.width, size.height));
              path.lineTo(size.width, nearLength!);
              vec.add(Offset(size.width, nearLength!));
            } else {
              path.lineTo(size.width - (nearLength! - size.height), size.height);
              vec.add(
                  Offset(size.width - (nearLength! - size.height), size.height));
            }
          }
          break;
        case RibbonLocation.bottomStart:
          path.moveTo(nearLength!, size.height);
          vec.add(Offset(nearLength!, size.height));
          path.lineTo(farLength!, size.height);
          vec.add(Offset(farLength!, size.height));
          if (farLength! <= size.height) {
            path.lineTo(0, size.height - farLength!);
            vec.add(Offset(0, size.height - farLength!));
            path.lineTo(0, size.height - nearLength!);
            vec.add(Offset(0, size.height - nearLength!));
          } else {
            path.lineTo(farLength! - size.height, 0);
            vec.add(Offset(farLength! - size.height, 0));
            if (nearLength! <= size.height) {
              path.lineTo(0, 0);
              vec.add(Offset(0, 0));
              path.lineTo(0, size.height - nearLength!);
              vec.add(Offset(0, size.height - nearLength!));
            } else {
              path.lineTo(nearLength! - size.height, 0);
              vec.add(Offset(nearLength! - size.height, 0));
            }
          }
          break;
        case RibbonLocation.bottomEnd:
          path.moveTo(size.width - nearLength!, size.height);
          vec.add(Offset(size.width - nearLength!, size.height));
          path.lineTo(size.width - farLength!, size.height);
          vec.add(Offset(size.width - farLength!, size.height));
          if (farLength! <= size.height) {
            path.lineTo(size.width, size.height - farLength!);
            vec.add(Offset(size.width, size.height - farLength!));
            path.lineTo(size.width, size.height - nearLength!);
            vec.add(Offset(size.width, size.height - nearLength!));
          } else {
            path.lineTo(size.width - (farLength! - size.height), 0);
            vec.add(Offset(size.width - (farLength! - size.height), 0));
            if (nearLength! <= size.height) {
              path.lineTo(size.width, 0);
              vec.add(Offset(size.width, 0));
              path.lineTo(size.width, size.height - nearLength!);
              vec.add(Offset(size.width, size.height - nearLength!));
            } else {
              path.lineTo(size.width - (nearLength! - size.height), 0);
              vec.add(Offset(size.width - (nearLength! - size.height), 0));
            }
          }
          break;
      }
    }
    path.close();
    List<Offset> vec2 = vec.toSet().toList();
    offsetRibbon = _center(vec2);
    // print('cx = ${offsetRibbon.dx},cy = ${offsetRibbon.dy}');
    return path;
  }

  double get _rotation {
    switch (location) {
      case RibbonLocation.topStart:
        return -math.pi / 4;
      case RibbonLocation.topEnd:
        return math.pi / 4;
      case RibbonLocation.bottomStart:
        return math.pi / 4;
      case RibbonLocation.bottomEnd:
        return -math.pi / 4;
    }
    return 0;
  }

  Offset _center(List<Offset> vecs) {
    double sum_x = 0, sum_y = 0, sum_s = 0;
    double x1 = vecs[0].dx;
    double y1 = vecs[0].dy;
    double x2 = vecs[1].dx;
    double y2 = vecs[1].dy;
    double x3, y3;
    for (int i = 2; i < vecs.length; i++) {
      x3 = vecs[i].dx;
      y3 = vecs[i].dy;
      double s = ((x2 - x1) * (y3 - y1) - (x3 - x1) * (y2 - y1)) / 2.0;
      sum_x += (x1 + x2 + x3) * s;
      sum_y += (y1 + y2 + y3) * s;
      sum_s += s;
      x2 = x3;
      y2 = y3;
    }
    double cx = sum_x / sum_s / 3.0;
    double cy = sum_y / sum_s / 3.0;
    return Offset(cx, cy);
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:ribbon_safe/ribbon_safe.dart';

void main() {
  runApp(MaterialApp(
    title: "ribbon demo",
    theme: ThemeData.light(),
    home: DemoPage(),
    // debugShowMaterialGrid:true,
    debugShowCheckedModeBanner: false,
  ));
}

class DemoPage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => _DemoPageState();
}

class _DemoPageState extends State<DemoPage> {
  double nearLength = 80;
  double farLength = 120;
  Color color = Colors.redAccent;
  RibbonLocation location = RibbonLocation.topStart;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Ribbon Demo'),
      ),
      body: Center(
        child: Column(mainAxisAlignment: MainAxisAlignment.center,
            // crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Ribbon(
                nearLength: nearLength,
                farLength: farLength,
                title: 'New!',
                titleStyle: TextStyle(
                    color: Colors.greenAccent,
                    fontSize: 18,
                    fontWeight: FontWeight.bold),
                color: Colors.redAccent,
                location: location,
                child: Container(
                    width: 300,
                    height: 160,
                    color:Colors.blueAccent,
                    child: Center(
                        child: Text(
                      'hello ribbon',
                      style: TextStyle(color: Colors.white, fontSize: 20),
                    ))),
              ),
              SizedBox(
                height: 16,
              ),
              Text(
                'RibbonLocation:',
                style: TextStyle(color: Colors.black),
              ),
              Row(
                mainAxisSize: MainAxisSize.min,
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  _buildRadio('topStart', RibbonLocation.topStart),
                  _buildRadio('topEnd', RibbonLocation.topEnd),
                ],
              ),
              Row(
                mainAxisSize: MainAxisSize.min,
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  _buildRadio('bottomStart', RibbonLocation.bottomStart),
                  _buildRadio('bottomEnd', RibbonLocation.bottomEnd),
                ],
              ),
              Row(
                mainAxisSize: MainAxisSize.min,
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  Text('nearLength:'),
                  Slider(
                    value: nearLength,
                    onChanged: (double value) {
                      setState(() {
                        nearLength = value;
                      });
                    },
                    label: 'nearLength',
                    min: 0,
                    max: 280,
                  ),
                ],
              ),
              Row(
                mainAxisSize: MainAxisSize.min,
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  Text('farLength:'),
                  Slider(
                    value: farLength,
                    onChanged: (double value) {
                      setState(() {
                        farLength = value;
                      });
                    },
                    label: 'nearLength',
                    min: 20,
                    max: 300,
                  ),
                ],
              ),
            ]),
      ),
    );
  }

  _buildRadio(String text, RibbonLocation value) {
    return Row(
      children: <Widget>[
        Text(text),
        Radio<RibbonLocation>(
            value: value,
            groupValue: location,
            onChanged: (RibbonLocation? value) {
              setState(() {
                location = value!;
              });
            }),
      ],
    );
  }
}
